game-core/src/main/java/games/strategy/net/AbstractServerMessenger.java

```diff
@@ -260,32 +212,60 @@ private void scheduleMacUnmuteAt(final String mac, final Instant checkTime) {
       cachedMacAddresses.put(uniquePlayerName, mac);
-      if (isLobby()) {
-        final String realName = IServerMessenger.getRealName(uniquePlayerName);
-        if (!liveMutedUsernames.contains(realName)) {
-          final Optional<Instant> muteTill = new MutedUsernameController(database).getUsernameUnmuteTime(realName);
-          muteTill.ifPresent(instant -> {
-            if (instant.isAfter(Instant.now())) {
-              // Signal the player as muted
-              liveMutedUsernames.add(realName);
-              scheduleUsernameUnmuteAt(realName, instant);
-            }
-          });
-        }
-        if (!liveMutedMacAddresses.contains(mac)) {
-          final Optional<Instant> muteTill = new MutedMacController(database).getMacUnmuteTime(mac);
-          muteTill.ifPresent(instant -> {
-            if (instant.isAfter(Instant.now())) {
-              // Signal the player as muted
-              liveMutedMacAddresses.add(mac);
-              scheduleMacUnmuteAt(mac, instant);
-            }
-          });
-        }
+      final String realName = IServerMessenger.getRealName(uniquePlayerName);
+      if (!liveMutedUsernames.contains(realName)) {
+        final Optional<Instant> muteTill = getUsernameUnmuteTime(realName);
+        muteTill.ifPresent(instant -> {
+          if (instant.isAfter(Instant.now())) {
+            // Signal the player as muted
+            liveMutedUsernames.add(realName);
+            scheduleUsernameUnmuteAt(realName, instant);
+          }
+        });
+      }
+      if (!liveMutedMacAddresses.contains(mac)) {
+        final Optional<Instant> muteTill = getMacUnmuteTime(mac);
+        muteTill.ifPresent(instant -> {
+          if (instant.isAfter(Instant.now())) {
+            // Signal the player as muted
+            liveMutedMacAddresses.add(mac);
+            scheduleMacUnmuteAt(mac, instant);
+          }
+        });
       }
     }
   }
 
+  /**
+   * Returns the instant at which the user associated with the specified username is to be unmuted or empty if the user
+   * is not currently muted.
+   *
+   * <p>
+   * Subclasses may override and are not required to call the superclass implementation. This implementation returns an
+   * empty instant indicating the user is not currently muted.
+   * </p>
+   *
+   * @param username The username of the user.
+   */
+  protected Optional<Instant> getUsernameUnmuteTime(final String username) {
+    return Optional.empty();
+  }
+
+  /**
+   * Returns the instant at which the user associated with the specified MAC is to be unmuted or empty if the user is
+   * not currently muted.
+   *
+   * <p>
+   * Subclasses may override and are not required to call the superclass implementation. This implementation returns an
+   * empty instant indicating the user is not currently muted.
+   * </p>
+   *
+   * @param mac The MAC of the user.
+   */
+  protected Optional<Instant> getMacUnmuteTime(final String mac) {
+    return Optional.empty();
+  }
```

Comment:
If this kind of logic is only being used by the GameServerMessenger: Shouldn't only this specific subclass implement it like this, or did you do it this way to save code for the TestServerMessenger instance?

Reply:
I actually tried that first.

I added a _Template Method_ that was called from `notifyPlayerLogin()` that contained the relevant code for the lobby and was empty for all other subclasses.  However, that required exposing a lot of private state of `AbstractServerMessenger` as protected due to how synchronization is implemented.  In the end, I went with the current solution because it keeps the state of `AbstractServerMessenger` nicely encapsulated.

In reality, all of this chat functionality should not be part of `AbstractServerMessenger` and its subclasses.  Rather, it should be mixed-in via a _Strategy_ or something.  However, that goes way beyond the purpose of this PR (to allow the rest of the lobby server code to be moved to the `lobby` project).